lib/shared/bucketing-assembly-script/assembly/eventQueue/eventQueue.ts

```diff
@@ -0,0 +1,197 @@
+     * It will generate an array of payloads, with each payload up to a maximum batch size
+     * (should be configurable by an option, nodejs uses 100).
+     *
+     * It should take the data from the eventQueue and aggregate events and generate request payloads.
+     * After generating these paylods it should clear the data from the eventQueue and aggregate events queues.
+     * The payload objects should be stored in a sending paylods queue with their status of 'inProgress'.
+     * Each Payload object should generate a unique payloadId that will be used for the onPayloadSuccess() and
+     * onPayloadFailure() methods.
+     *
+     * If onPayloadSuccess() is called for a payloadId, it can be removed from the sending payloads queue.
+     *
+     * If onPayloadFailure() is called for a payloadId and retryable = true, it should be marked as 'failed'.
+     * If it retryable = false, it should be removed from the sending payload queue.
+     * Then on subsequent calls to `flush()` the 'failed' payloads should be marked as 'inProgress' and returned
+     * as payloads to be sent by the Native code again.
+     *
+     * If there are payloads that are still marked as `inProgress` when `flush()` is called, it should return an
+     * exception to the native code.
+     */
+    flush(): string {
+        const payloads: FlushPayload[] = this.requestPayloadManager.constructFlushPayloads(
+            this.userEventQueue,
+            this.aggEventQueue
+        )
+        this.userEventQueue = new Map<string, UserEventsBatchRecord>()
+        this.aggEventQueue = new Map<string, Map<string, Map<string, Map<string, i64>>>>()
+        return jsonArrFromValueArray(payloads).stringify()
+    }
+
+    onPayloadSuccess(payloadId: string): void {
+        this.requestPayloadManager.markPayloadSuccess(payloadId)
+    }
+
+    onPayloadFailure(payloadId: string, retryable: boolean): void {
+        this.requestPayloadManager.markPayloadFailure(payloadId, retryable)
+    }
+
+    queueEvent(user: DVCPopulatedUser, event: DVCEvent, featureVariationMap: Map<string, string>): void {
+        // TODO: Implement max queue size
+        // this.maxEventQueueSize = bucketedConfig?.project.settings.sdkSettings?.eventQueueLimit
+        // ?? this.maxEventQueueSize
+
+        const requestEvent = new DVCRequestEvent(event, user.user_id, featureVariationMap)
+        this.addEventToQueue(user, requestEvent)
+    }
+
+    private addEventToQueue(user: DVCPopulatedUser, event: DVCRequestEvent): void {
+        // TODO: implement?
+        // if (this.eventQueueSize() >= this.maxEventQueueSize) {
+        //     this.logger.warn(`Max event queue size reached, dropping event: ${event}`)
+        //     this.flushEvents()
+        //     return
+        // }
+
+        const user_id = user.user_id
+        let userEvents: UserEventsBatchRecord
+        if (!this.userEventQueue.has(user_id)) {
```

Comment:
grouping the events purely by user_id means that you'll be using the user data as specified by the first event that was queued, even if the user data changes later on (but keeps the same user id). Shouldn't you group them by the whole user data object to make sure the data is correct on each event?

Reply:
the user is updated to the latest user in line `106` below